Apollo types

AlexanderJernstrom opened the pull request.

AElmoznino reviewed (changes requested):
Good job!  👌 
Could you please add a step-by-step section (in English) on how to generate and use the introspection? Also, do you have to set the URL to GraphCMS somewhere for the introspection, or how is it done?

AlexanderJernstrom commented:
0. Install the apollo CLI globally: `npm i  -g apollo`.
1. Introspect the schema and download a schema.json file with the following command: `apollo client:download-schema --endpoint=GRAPH_CMS_ENDPOINT schema.json`.
2. Create a .ts file with your query or mutation in the graphql folder. In the typescript file make sure that you are exporting your query or mutation.
3. I have provided a codegen command in the package.json. Run `npm run codegen`.
4. Now you can use your types for the desired query or mutation anywhere in the app. The types are are located in folder called --generated-- in src/graphql.

AElmoznino commented:
@AlexanderJernstrom I forgot to mention - could you add it to the README (in this PR)?

AElmoznino commented on README.md:
```diff
@@ -30,6 +30,16 @@ To get more help on the Angular CLI use `ng help` or go check out the [Angular C
 
+## Generate types for apollo
+
+If this is the first time you are generating types follow all the steps, but if you already have the apollo installed and downloaded the schema feel free to skit to step number two.
```
```suggestion
If this is the first time you are generating types follow all the steps, but if you already have Apollo installed and downloaded the schema, feel free to skip to step number two.
```

AElmoznino reviewed (changes requested):
I notice that the tests (`npm run test`) are failing with the following error message:
```
ERROR in src/app/modules/info/info.component.spec.ts(3,25): error TS2305: Module '"./info.component"' has no exported member 'GET_PAGE'.
src/app/modules/lesson/lesson.component.spec.ts(3,27): error TS2305: Module '"./lesson.component"' has no exported member 'GET_LESSON'.
src/app/modules/overview/overview.component.spec.ts(3,29): error TS2305: Module '"./overview.component"' has no exported member 'GET_OVERVIEW'.
src/app/modules/start/start.component.spec.ts(3,26): error TS2305: Module '"./start.component"' has no exported member 'GET_DIFFICULTIES'.
```

I think you know how to fix it, but let me know if you need help :)

AElmoznino commented:
Also, I tried running step 1 and it gives me a different output in `schema.json` , mostly adding this in a ton of places:
```
     "isDeprecated": false,
     "deprecationReason": null
```
I assume GraphCMS has updated something since you ran the command? 🤔

AlexanderJernstrom commented:
I fixed the issues with the test files importing the queries the wrong way but now I am getting a new type of error: 
`
LessonComponent with response renders a header when it has received data FAILED
`
should I get this error or have I done somethign wrong?

AlexanderJernstrom commented:
I saw now that you actually don't have to download the schema again because I added it in the commit. I think "isDeprecated" and "deprecationReason" properties were there the first time I generated the schema.

AElmoznino commented:
@AlexanderJernstrom you haven't done anything wrong :) 
I've [opened a PR](https://github.com/AElmoznino/coderdojo-web-angular/pull/14) (didn't have time to finish it now) to upgrade to Angular 11, but now most of the tests are failing. I'd like to fix that before merging it. That PR will also conflict with this one (in package.json and package-lock.json).

Also, I think isDeprecated and deprecationReason should be checked in, right?

AlexanderJernstrom commented:
I think the isDeprecated has somethign to do wiht the fact that in graphql you can set a field as deprecated, as mentioned in this stackoverflow post https://stackoverflow.com/questions/49175623/marking-fields-as-deprecated-with-graphql-shorthand.

AElmoznino commented:
1) My PR is now merged and all the tests are passing 🎉 Feel free to rebase or merge master into this one.
2) If me or someone else tries to run the codegen there will be a discrepancy in the generated files (since isDeprecated will  be added), so better commit that here too,  so it doesn't feel "automagic" to the next person.

AlexanderJernstrom commented:
The deprecationReason and isDeprecated property only lives inside the schema.json file. It will not transfer over to the generated types. The types look like this.
![Annotation 2021-03-26 143907](https://user-images.githubusercontent.com/46424392/112639939-233ab300-8e41-11eb-9a66-e8b28923ea4e.png)

AElmoznino commented:
@AlexanderJernstrom ok, cool! So I guess the `schema.json` should be added to the .gitignore file, if it's not relevant for all  users to have the same version of it?

AElmoznino commented:
I believe you need to merge `master` into this branch, or rebase this branch, in order to resolve the merge conflicts.
Do you know how to do that, or do you need help?

AlexanderJernstrom commented:
I added the schema.json to the .gitignore and then I rebased, when I committed the changes I saw that the netlify tests had failed. It was something about a package called node-sass, so I will have to fix that before I merge the changes.
